### PR TITLE
[iOS] [4.0] iOS Plugins

### PR DESCRIPTION
**Pull request by naithar**
Implements https://github.com/godotengine/godot-proposals/issues/1185 for `master` branch.
[Godot iOS Plugin template](https://github.com/naithar/godot_ios_plugin)
*Since this plugin implementation directly uses Godot's source headers it's important to use correct compilation flags.*

This PR allows iOS project to load plugins at runtime. 
This would allow developers to work on platform functionality (like Facebook Auth, AdMob and other platform specific modules) without requiring to rebuild an engine to embed it into application. 
~~Godot's iOS binary would probably be needed to be supplied like Androids' AAR library is.~~ Compiling plugins without Godot's iOS binary actually works fine, only headers are required, but correct compilation flags should be used.
Plugins is also more portable, even than `GDNative` which requires additional setup for same functionality.
Plugin also produces smaller binary size for same code - ~800KB for GameCenter plugin vs ~4MB for GameCenter in `GDNative` as `framework` or ~800 KB for module + ~3MB for `GDNative` itself as dependency. 

Also plugins can work with current module system, it's just makes things a lot easier for iOS. So it should be possible to migrate some parts of this to Godot 3.

iOS Plugin is basically a `.gdip` (like `.gdap` for Android) configuration file with static library (which is linked to Godot's binary) and dependencies bundled with it. 
`.gdip` file format allows to specify dependency libraries, frameworks and resource files as well as `plist` values and system capabilities. Dependencies as well as binary itself could use absolute and relative path. 
Plugin files should be placed in `ios/plugins` folder or it's child folder.
Supports multiple target binaries.

Some parts for implementation was taken from Android's plugin system. Such as configuration file parsing, configuration observer.

<details>
<summary>
Plugins configuration syntax
</summary>

```
[config]
name=""
binary=""

initialization=""
deinitialization=""

[dependencies]
linked=[]
embedded=[]
system=[]

capabilities=[]

files=[]

[plist]
```

</details>

<details>

<summary>
Example
</summary>

Having project structure like this:
```
*project
|-...
|--ios
  |--plugins
    |--gamecenter
      |--gamecenter.gdip
      |--gamecenter_lib.release.a
      |--gamecenter_lib.debug.a
    |--arkit
      |--arkit.gdip
      |--arkit_lib.release.a
      |--arkit_lib.debug.a
    |--camera
      |--camera.gdip
      |--camera_lib.a
    |--inappstore
      |--inappstore.gdip
      |--inappstore_lib.a
    |--icloud
      |--icloud.gdip
      |--icloud_lib.a
```

Enabled this export options in editor:
Снимок экрана 2020-08-14 в 20 43 39

`dymmy.cpp` file in Xcode project looks like this when exporting multiple plugins.

```
// Godot Plugins
void godot_ios_plugins_initialize();
void godot_ios_plugins_deinitialize();
// Exported Plugins

// Plugin: GameCenter
extern void register_gamecenter_types();
extern void unregister_gamecenter_types();

// Plugin: Camera
extern void register_camera_types();
extern void unregister_camera_types();

// Plugin: InAppStore
extern void register_inappstore_types();
extern void unregister_inappstore_types();

// Plugin: ARKit
extern void register_arkit_types();
extern void unregister_arkit_types();

// Plugin: iCloud
extern void register_icloud_types();
extern void unregister_icloud_types();

// Use Plugins
void godot_ios_plugins_initialize() {
	register_gamecenter_types();
	register_camera_types();
	register_inappstore_types();
	register_arkit_types();
	register_icloud_types();
}

void godot_ios_plugins_deinitialize() {
	unregister_gamecenter_types();
	unregister_camera_types();
	unregister_inappstore_types();
	unregister_arkit_types();
	unregister_icloud_types();
}
```

This gives developer access to `GameCenter`, `InAppStore`, `iCloud` and other functionality for iOS that previously required rebuilding an engine to be enabled or disabled.

</details>

<details>

<summary>
Obj-C interaction
</summary>

Modifying template likes this:
```
#import <Foundation/Foundation.h>

#include "core/project_settings.h"
#include "core/class_db.h"
#import "platform/iphone/app_delegate.h"
#import "platform/iphone/view_controller.h"

#import "godot_plugin_implementation.h"

void PluginExample::_bind_methods() {
    ClassDB::bind_method(D_METHOD("example_method"), &PluginExample::example_method);
}

void PluginExample::example_method(String str) {
    NSLog(@"foo");
    print_error(str);
    NSLog(@"%@", AppDelegate.viewController);
    
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(100, 100, 300, 300)];
    view.backgroundColor = [UIColor redColor];
    
    [AppDelegate.viewController.view addSubview:view];
}
```

Results in this:

So direct communication with ObjC is not a problem.

</details>

<details>

<summary>
Signals integration
</summary>

```

static int i = 0;

void PluginExample::_bind_methods() {
    ClassDB::bind_method(D_METHOD("foo"), &PluginExample::foo);
    ADD_SIGNAL(MethodInfo("event", PropertyInfo(Variant::INT, "id")));
}

Error PluginExample::foo() {
    NSLog(@"foo");
    
    i += 1;
    emit_signal("event", i);
    
    return OK;
}
```

</details>

### CODE REVIEW

**Comment by naithar** [comment]
Added a [iOS template](https://github.com/naithar/godot_ios_plugin) for plugins. 
Currently it seems that plugin doesn't really need to actually link to Godot's iOS binary. Only headers are required including the ones generated by compiling.

**Comment by naithar** [comment]
I've retested and everything seems to work fine.

But @cagdasc is also implementing his own plugin system in #41269, so the best approach could be chosen :)

**Comment by naithar** [comment]
So I seems that I've got an issue with current implementation.
Exporting [template](https://github.com/naithar/godot_ios_plugin) and closing the app results in crash in `ClassDB::cleanup` method at `memdelete(ti.method_map[*m]);` with logs 
```
godot_app(38015,0x1054c9880) malloc: *** error for object 0x2826e5660: pointer being freed was not allocated
godot_app(38015,0x1054c9880) malloc: *** set a breakpoint in malloc_error_break to debug
```

Backtrace:

```
frame #0: 0x000000019e6bedf0 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000019e5de930 libsystem_pthread.dylib`pthread_kill + 228
    frame #2: 0x000000019e56cba4 libsystem_c.dylib`abort + 104
    frame #3: 0x000000019e5cffdc libsystem_malloc.dylib`malloc_vreport + 564
    frame #4: 0x000000019e5d01a4 libsystem_malloc.dylib`malloc_report + 64
    frame #5: 0x000000019e5c3d1c libsystem_malloc.dylib`free + 436
    frame #6: 0x000000010292ed80 godot_app`void memdelete<MethodBind>(p_class=0x00000002826e5660) at memory.h:115:13
  * frame #7: 0x00000001028952b4 godot_app`ClassDB::cleanup() at class_db.cpp:1553:4
    frame #8: 0x0000000102967ebc godot_app`unregister_core_types() at register_core_types.cpp:320:2
    frame #9: 0x0000000100e5d568 godot_app`Main::cleanup() at main.cpp:2526:2
    frame #10: 0x0000000100e3ebf0 godot_app`iphone_finish() at godot_iphone.mm:128:2
    frame #11: 0x0000000100e40fc4 godot_app`-[AppDelegate applicationWillTerminate:](self=0x0000000280fc4770, _cmd="applicationWillTerminate:", application=0x000000011df04950) at app_delegate.mm:121:2
```

This issue is not specific to template - if I build using templates `SConstruct` this error occurs.

**Comment by naithar** [comment]
> So it seems that I've got an issue with current implementation.
> Exporting [template plugin](https://github.com/naithar/godot_ios_plugin) and closing the app results in crash in `ClassDB::cleanup` method at `memdelete(ti.method_map[*m]);` with logs
> 
> ```
> godot_app(38015,0x1054c9880) malloc: *** error for object 0x2826e5660: pointer being freed was not allocated
> godot_app(38015,0x1054c9880) malloc: *** set a breakpoint in malloc_error_break to debug
> ```
> 
> Backtrace:
> 
> ```
> frame #0: 0x000000019e6bedf0 libsystem_kernel.dylib`__pthread_kill + 8
>     frame #1: 0x000000019e5de930 libsystem_pthread.dylib`pthread_kill + 228
>     frame #2: 0x000000019e56cba4 libsystem_c.dylib`abort + 104
>     frame #3: 0x000000019e5cffdc libsystem_malloc.dylib`malloc_vreport + 564
>     frame #4: 0x000000019e5d01a4 libsystem_malloc.dylib`malloc_report + 64
>     frame #5: 0x000000019e5c3d1c libsystem_malloc.dylib`free + 436
>     frame #6: 0x000000010292ed80 godot_app`void memdelete<MethodBind>(p_class=0x00000002826e5660) at memory.h:115:13
>   * frame #7: 0x00000001028952b4 godot_app`ClassDB::cleanup() at class_db.cpp:1553:4
>     frame #8: 0x0000000102967ebc godot_app`unregister_core_types() at register_core_types.cpp:320:2
>     frame #9: 0x0000000100e5d568 godot_app`Main::cleanup() at main.cpp:2526:2
>     frame #10: 0x0000000100e3ebf0 godot_app`iphone_finish() at godot_iphone.mm:128:2
>     frame #11: 0x0000000100e40fc4 godot_app`-[AppDelegate applicationWillTerminate:](self=0x0000000280fc4770, _cmd="applicationWillTerminate:", application=0x000000011df04950) at app_delegate.mm:121:2
> ```
> 
> This issue is not specific to template - if I build any iOS module using template's `SConstruct` this error occurs.
> 
> Same issue also appeared when I was implementing Vulkan support for iOS: [link](https://github.com/godotengine/godot/pull/40434#issuecomment-658996527). But crash was happening for specific(?) combination of output/input parameters and it stopped after rebase from master.
> This time it happens for every method.
> 
> Removing `-DTYPED_METHOD_BIND` flag removed crash for methods with no output/input parameters, but app is still crashing for anything else.
> 
> **Edit**
> 
> Crash happens at line `155: p_class->~T();` of `void memdelete<MethodBind>(MethodBind*)` method.
> 
> **Edit:**
> 
> ~Moving Plugin code into `modules` folder in Godot's source and building it there, seems to be fixing the issue even without using new Godot's binary - using newly built plugin `.a` library is enough for everything to work correctly.~
> ~Not really working with old Godot's binary that doesn't contain changes in PR.~
> 
> Seems like the problem is with compilation flags used in example project.

Adding a `PTRCALL_ENABLED` flag for plugin compilation fixed the crash

**Comment by naithar** [comment]
Retested with every module being rebuilt with new `SConstruct` in templates. No crashes on cleanup and everything seems to be working correctly.

**Comment by m4gr3d** [comment]
cc @BastiaanOlij since this may affect how ARKit function is provided.

**Comment by naithar** [comment]
Is there any ARKit test project for `4.0` that I can use? 
I've just tested ARKit for `3.2` and made some changes for it to work, which I'll push in #41340

**Comment by tolgakaranlik** [comment]
naithar I believe what you have done so far is amazing. My humble request from you is that please consider makin an example function which takes a parameter and emit an example signal in your plugin template if you can find some spare time. It would be very useful for guys like me =) Anyway, you have my gratitude, keep up your good work!

**Comment by naithar** [comment]
@tolgakaranlik for now you can probably look at `websocket` module for reference

File: modules/SCsub [hunk 1]
```diff
@@ -47,7 +47,14 @@ for name, path in env.module_list.items():
...
     # on iOS, so we handle those specially.
-    if env["platform"] == "iphone" and name in ["arkit", "camera"]:
+    if env["platform"] == "iphone" and name in [
+        "arkit",
+        "camera",
+        "camera_iphone",
+        "gamecenter",
+        "inappstore",
+        "icloud",
+    ]:
```
[the remark this answers is not in the document]
Comment (naithar) [reply modules/SCsub:57]: Yeah, it should :) This way plugins would be probably easier to fix and modify. 
I think single repo for all godot's plugins should be enough.

**Review by bruvzg — APPROVED** [review]
Why splitting `camera_iphone` from `camera`? It's almost the same as macOS code, I would merge macOS and iOS code instead.

And instead of renaming `register_types.*` to `camera_module.*` it's probably better to add a function to the `config.py` to prevent it from adding it `register_module_types.gen.cpp` on iOS.

**Comment by naithar** [comment]
Well, the reason is to simplify the future split/migration of this modules from Godot's source to separate repo without breaking how modules work in the mean time. It's just seemed easier to do it this way, without introducing additional configuration that would require additional testing/fixing.

**Comment by BastiaanOlij** [comment]
Haven't dived into the code yet but initial reaction on whats written about this, this is really cool. I'm even wondering if we shouldn't investigate moving to this as an approach for all platforms with Godot 4 as a replacement of GDNative.

Bit more tricky because it basically means we'd need to ship a linker with Godot to link libraries together with the Godot core to produce the final executable, but it would result in us being able to properly subclass just like we do in modules, instead of the bindings and wrapper approach in GDNative which has plenty of issues.

On iOS we're statically linking anyway due to Apple not allowing dynamic libraries to begin with and both iOS and Android already require Xcode and Android studio respectively to do the deployment so...

Only other on topic remark I have at this time is that this would be ideal for splitting off ARKit into a plugin, but I don't get the benefit of splitting off the camera logic, it doesn't link in any additional libraries and there is hardly any code that we remove from the core that could be considered bloat

**Comment by BastiaanOlij** [comment]
Oh also note that if I understand this correctly, it's not just the compiler settings that need to match, you have to version match your plugin exactly to Godot. As class definitions change from release to release your code could be assuming a different class structure to what is actually used.

**Comment by naithar** [comment]
> Haven't dived into the code yet but initial reaction on whats written about this, this is really cool. I'm even wondering if we shouldn't investigate moving to this as an approach for all platforms with Godot 4 as a replacement of GDNative.

macOS could probably use same approach, but it would require making a Xcode template for it. 
I'm planning some changes for macOS, like ARC and maybe something else, but changing template is a different thing. @bruvzg could probably tell if it's possible.
Also GDNative allow almost any language to be used for Godot scripting, so I'm not sure if it should be completely replaced.

> On iOS we're statically linking anyway due to Apple not allowing dynamic libraries to begin with and both iOS and Android already require Xcode and Android studio respectively to do the deployment so...

iOS does allow dynamic libraries, but requires iOS 8 for it and it should be bundled as `.framework` or `.xcframework`. GDNative for iOS already supports it (https://github.com/godotengine/godot/pull/39996 in case you missed it). 

> I don't get the benefit of splitting off the camera logic, it doesn't link in any additional libraries and there is hardly any code that we remove from the core that could be considered bloat

It should be pretty easy to revert it back to module if it's needed. 
But I think platform specific code should be moved to plugins, especially if camera would require new functionality in future.

> Oh also note that if I understand this correctly, it's not just the compiler settings that need to match, you have to version match your plugin exactly to Godot. As class definitions change from release to release your code could be assuming a different class structure to what is actually used.

Yeah, both compiler flags and headers should match, sometimes it works without up-to-date headers, but it's not ideal case. 
I think I've mentioned somewhere (not sure where) that it would be nice to have headers available for download just like `.aar` library for Android is for release builds. It should solve most of the issues with getting and generating them from source files.

**Comment by bruvzg** [comment]
>  without breaking how modules work in the mean time

4.0 is breaking compatibility in many places, this is not a concern.

> Bit more tricky because it basically means we'd need to ship a linker with Godot to link libraries together with the Godot core to produce the final executable, but it would result in us being able to properly subclass just like we do in modules, instead of the bindings and wrapper approach in GDNative which has plenty of issues.

> macOS could probably use same approach, but it would require making a Xcode template for it.

It's possible, but we want to support exporting for macOS from other platforms, dependence on Xcode is not an option.

Using bundled linker, maybe. LLD do support all required platforms and architectures, and AFAIK can be used as standalone executable, but it's almost 50 MB in size (at least on macOS).

And, there're plans to compile GDScript to native code, so it's probably going to happen anyway.

> But I think platform specific code should be moved to plugins, especially if camera would require new functionality in future.

It's my personal preference, but I would like to have modules, plugins and C++ GDNative completely interchangeable (as long as all three are supported), exactly same code, just different build options.

> Oh also note that if I understand this correctly, it's not just the compiler settings that need to match, you have to version match your plugin exactly to Godot. As class definitions change from release to release your code could be assuming a different class structure to what is actually used.

Also, in case of rebuilding (or downloading prebuild) only some plugins, minor compiler changes should not cause issues, but different compilers (GCC ↔ clang, GCC ↔ MSVC) definitely won't work (it works with GDNative which is pure-C API).

**Comment by naithar** [comment]
> 4.0 is breaking compatibility in many places, this is not a concern.

Same system is used for `3.2` PR so I wouldn't break it for now. And migration to a separate repo would remove this modules anyway, so I'm not sure it's really needed.

> It's my personal preference, but I would like to have modules, plugins and C++ GDNative completely interchangeable (as long as all three are supported), exactly same code, just different build options.

This plugin system allows it to some extend. Not sure about GDNative, but module/plugin code is exactly the same.

**Comment by naithar** [comment]
> Using bundled linker, maybe. LLD do support all required platforms and architectures, and AFAIK can be used as standalone executable, but it's almost 50 MB in size (at least on macOS).
> 
> And, there're plans to compile GDScript to native code, so it's probably going to happen anyway.

Well it could come as editor (?) plugin in case user doesn't need this feature, but I'm not really sure how it would work for now.

**Comment by akien-mga** [comment]
Needs a rebase, otherwise I guess we could merge the current state already and build upon it if further changes are needed?

**Comment by naithar** [comment]
Rebased.
I'll try to run some more tests, but it should be save to merge.

> I guess we could merge the current state already and build upon it if further changes are needed

Yeah, I think so. Moving modules to separate repo could be done in next PRs. 
Also implementing a pluggable `AppDelegate` to simplify support for Push Notifications, Facebook or any other SDK that requires swizzling right now seems like a completely separate functionality.

**Comment by akien-mga** [comment]
Thanks a ton for your work!

**Comment by alexzheng** [comment]
A little suggestion:
Make the godot_ios_plugins_initialize called before any GDScripts.
It's strange that a plugin is available in the _ready method and not available in _init.

Take the Game Center plugin for example
add an autoload node, the scripts for that node is:

func _init() -> void:
	if Engine.has_singleton("GameCenter"):
		printt("has_singleton GameCenter in _init")
	else:
		printt("not has_singleton GameCenter _init")

func _ready() -> void:
	if Engine.has_singleton("GameCenter"):
		printt("has_singleton GameCenter in _ready")
	else:
		printt("not has_singleton GameCenter in _ready")


The results is  
not has_singleton GameCenter _init
has_singleton GameCenter in _ready

Before  the _init is happened before the godot_ios_plugins_initialize called.

**Comment by naithar** [comment]
@alexzheng this two branches should probably allow this. But they require testing.
https://github.com/naithar/godot/tree/fix/ios-plugin-initialization
https://github.com/naithar/godot/tree/fix/ios-plugin-initialization-4.0
If you can test `3.2` version it would help greatly.

**Comment by alexzheng** [comment]
I have tested in 3.2.1 rc 1
the result is it does not work.

output in Xcode:
**SCRIPT ERROR**: Compile Error: Can't load global class GameCenter, cyclic reference?

My steps:
build libs from the Godot source
replace the libgodot.iphone.release.fat.a and libgodot.iphone.debug.fat.a in iphone.zip in the templates
export an Xcode project from godot editor.

it show the above error.

The process is not very straightforward and I'm not sure if my steps is correct.

So I did the above step again for the source code of 3.2.4 rc1,  it can run without error.

**Comment by naithar** [comment]
RC1 doesn't have changes from https://github.com/naithar/godot/tree/fix/ios-plugin-initialization in it and it wouldn't until it's tested.
You have to manually build iOS template library from source. After that you can simply change `<project_name>.a` in exported Xcode project with the library you've just built.

> output in Xcode:
> SCRIPT ERROR: Compile Error: Can't load global class GameCenter, cyclic reference?

Without example I cannot reproduce this. 

With the code that you've provided earlier there are no errors:
```
func _init() -> void:
if Engine.has_singleton("GameCenter"):
printt("has_singleton GameCenter in _init")
else:
printt("not has_singleton GameCenter _init")

func _ready() -> void:
if Engine.has_singleton("GameCenter"):
printt("has_singleton GameCenter in _ready")
else:
printt("not has_singleton GameCenter in _ready")
```

**Comment by alexzheng** [comment]
I made modification in RC1 with your committed code, just move godot_ios_plugins_initialize from set_main_loop to start

**Comment by naithar** [comment]
I can't reproduce your `SCRIPT ERROR: Compile Error: Can't load global class GameCenter, cyclic reference` issue with my project. You have to make MRP.

**Comment by alexzheng** [comment]
That's my custom gdscript class, it can run in RC1

class_name GameCenter

var _GameCenter = null


func _init() -> void:
	if Engine.has_singleton("GameCenter"):
		_GameCenter = Engine.get_singleton("GameCenter")
		_GameCenter.authenticate()
		printt("2 has_singleton GameCenter")
	else:
		printt("2 not has_singleton GameCenter")


When it run in your branch:
**SCRIPT ERROR**: Parse Error: The class "GameCenter" shadows a native class.
2021-02-01 21:29:24.666473+0800 Idiom[44623:10254439]    At: res://Idiom/autoload/gamekit.gdc:4:GDScript::load_byte_code() - Parse Error: The class "GameCenter" shadows a native class.
2021-02-01 21:29:24.666487+0800 Idiom[44623:10254439] **ERROR**: Method failed. Returning: ERR_PARSE_ERROR
2021-02-01 21:29:24.666498+0800 Idiom[44623:10254439]    At: modules/gdscript/gdscript.cpp:801:load_byte_code() - Method failed. Returning: ERR_PARSE_ERROR
2021-02-01 21:29:24.666508+0800 Idiom[44623:10254439] **ERROR**: Condition "err != OK" is true. Returned: RES()
2021-02-01 21:29:24.666521+0800 Idiom[44623:10254439]    At: modules/gdscript/gdscript.cpp:2311:load() - Condition "err != OK" is true. Returned: RES()
2021-02-01 21:29:24.666533+0800 Idiom[44623:10254439] **ERROR**: Condition "found" is true. Returned: RES()
2021-02-01 21:29:24.666543+0800 Idiom[44623:10254439]    At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
2021-02-01 21:29:24.666667+0800 Idiom[44623:10254439] **SCRIPT ERROR**: Compile Error: Can't load global class GameCenter, cyclic reference?
2021-02-01 21:29:24.666756+0800 Idiom[44623:10254439]    At: res://Idiom/autoload/global.gdc:50:GDScript::load_byte_code() - Compile Error: Can't load global class GameCenter, cyclic reference?
2021-02-01 21:29:24.666855+0800 Idiom[44623:10254439] **ERROR**: Method failed. Returning: ERR_COMPILATION_FAILED
2021-02-01 21:29:24.666927+0800 Idiom[44623:10254439]    At: modules/gdscript/gdscript.cpp:809:load_byte_code() - Method failed. Returning: ERR_COMPILATION_FAILED
2021-02-01 21:29:24.666951+0800 Idiom[44623:10254439] **ERROR**: Condition "err != OK" is true. Returned: RES()
2021-02-01 21:29:24.667074+0800 Idiom[44623:10254439]    At: modules/gdscript/gdscript.cpp:2311:load() - Condition "err != OK" is true. Returned: RES()
2021-02-01 21:29:24.667130+0800 Idiom[44623:10254439] **ERROR**: Condition "found" is true. Returned: RES()
2021-02-01 21:29:24.667232+0800 Idiom[44623:10254439]    At: core/io/resource_loader.cpp:279:_load() - Condition "found" is true. Returned: RES()
2021-02-01 21:29:24.667302+0800 Idiom[44623:10254439] **ERROR**: Condition "res.is_null()" is true. Continuing.
2021-02-01 21:29:24.667367+0800 Idiom[44623:10254439]    At: main/main.cpp:1762:start() - Condition "res.is_null()" is true. Continuing.

**Comment by alexzheng** [comment]
After change class_name GameCenter to class_name MyGameCenter, your branch works, and it output:
has_singleton GameCenter in _init, it works as expected.

**Comment by naithar** [comment]
Didn't you say that `_init` isn't working with RC1? So your script wasn't working in RC1 too?

Also `SCRIPT ERROR: Parse Error: The class "GameCenter" shadows a native class.` is pretty clear error in my opinion, just change the class name.

**Comment by alexzheng** [comment]
Yes, I did not notice my class name is the same as the native class in RC1.  However it can works in RC1,  only has the error that the Plugin GameCenter is not available in _init.

**Comment by naithar** [comment]
That's because `GameCenter` plugin isn't registered yet when `_init` is called. And that's probably enough to allow to shadow native class name.

**Comment by alexzheng** [comment]
Whatever,Your modification has fix that issue.

**Pull request merged**